### PR TITLE
Change Horizontal Pod Autoscaler to scale on OpenTelemetry Collector …

### PR DESCRIPTION
**Pull request by kevinearls**
…object

Signed-off-by: Kevin Earls <kearls@redhat.com>

### CODE REVIEW

File: bundle/manifests/opentelemetry.io_opentelemetrycollectors.yaml [hunk 1]
```diff
@@ -223,6 +223,11 @@ spec:
                   If MaxReplicas is set autoscaling is enabled.
                 format: int32
                 type: integer
+              minReplicas:
+                description: MinReplicas sets a lower bound to the autoscaling feature.  It
+                  must be at least 1
```
Comment (yuriolisa) [inline bundle/manifests/opentelemetry.io_opentelemetrycollectors.yaml:228]: Please remove this line break.

File: config/crd/bases/opentelemetry.io_opentelemetrycollectors.yaml [hunk 2]
```diff
@@ -221,6 +221,11 @@ spec:
                   If MaxReplicas is set autoscaling is enabled.
                 format: int32
                 type: integer
+              minReplicas:
+                description: MinReplicas sets a lower bound to the autoscaling feature.  It
```
Comment (yuriolisa) [inline config/crd/bases/opentelemetry.io_opentelemetrycollectors.yaml:225]: Please remove this line break.

**Review by yuriolisa — APPROVED** [review]
Only small code nit. LGTM.

File: bundle/manifests/opentelemetry.io_opentelemetrycollectors.yaml (hunk 1 above)
Comment (kevinearls) [reply bundle/manifests/opentelemetry.io_opentelemetrycollectors.yaml:228]: @yuriolisa Is that possible?  This is generated code based on this comment, which is all on one line.  https://github.com/open-telemetry/opentelemetry-operator/blob/98739669d581e64458cd887609198489022f756e/apis/v1alpha1/opentelemetrycollector_types.go#L40

Comment (yuriolisa) [reply bundle/manifests/opentelemetry.io_opentelemetrycollectors.yaml:228]: It's possible that I overlooked this code.

**Review by pavolloffay — CHANGES_REQUESTED** [review]

**Comment by kevinearls** [comment]
@pavolloffay Can you please review this again and then merge it if it looks ok?

File: apis/v1alpha1/opentelemetrycollector_types.go [hunk 3]
```diff
@@ -33,10 +33,14 @@ type OpenTelemetryCollectorSpec struct {
 	// +optional
 	Args map[string]string `json:"args,omitempty"`
 
-	// Replicas is the number of pod instances for the underlying OpenTelemetry Collector
+	// Replicas is the number of pod instances for the underlying OpenTelemetry Collector.  Set this if your are not using autoscaling
```
Comment (frzifus) [inline apis/v1alpha1/opentelemetrycollector_types.go:36]:
```suggestion
	// Replicas is the number of pod instances for the underlying OpenTelemetry Collector. Set this if your are not using autoscaling
```

File: config/crd/bases/opentelemetry.io_opentelemetrycollectors.yaml [hunk 4]
```diff
@@ -221,6 +221,11 @@ spec:
                   If MaxReplicas is set autoscaling is enabled.
                 format: int32
                 type: integer
+              minReplicas:
+                description: MinReplicas sets a lower bound to the autoscaling feature.  Set
```
Comment (frzifus) [inline config/crd/bases/opentelemetry.io_opentelemetrycollectors.yaml:225]:
```suggestion
                description: MinReplicas sets a lower bound to the autoscaling feature. Set
```

File: config/crd/bases/opentelemetry.io_opentelemetrycollectors.yaml [hunk 5]
```diff
@@ -221,6 +221,11 @@ spec:
...
+                  this if your are using autoscaling.  It must be at least 1
```
Comment (frzifus) [inline config/crd/bases/opentelemetry.io_opentelemetrycollectors.yaml:226]:
```suggestion
                  this if your are using autoscaling. It must be at least 1
```

File: config/crd/bases/opentelemetry.io_opentelemetrycollectors.yaml [hunk 6]
```diff
@@ -473,7 +478,7 @@ spec:
                 x-kubernetes-list-type: atomic
               replicas:
                 description: Replicas is the number of pod instances for the underlying
-                  OpenTelemetry Collector
+                  OpenTelemetry Collector.  Set this if your are not using autoscaling
```
Comment (frzifus) [inline config/crd/bases/opentelemetry.io_opentelemetrycollectors.yaml:481]:
```suggestion
                  OpenTelemetry Collector. Set this if your are not using autoscaling
```

File: hack/install-metrics-server.sh [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+# Install metrics-server on kind clusters for autoscale tests.  Note: This is not needed for minikube,
+# you can just add --addons "metrics-server" to the start command.
+kubectl apply -f https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/components.yaml
+kubectl patch deployment -n kube-system metrics-server --type "json" -p '[{"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value": --kubelet-insecure-tls}]'
+kubectl wait --for=condition=available deployment/metrics-server -n kube-system  --timeout=5m
```
Comment (frzifus) [inline hack/install-metrics-server.sh:7]: missing CRLF

File: tests/e2e/autoscale/00-install.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+    exporters:
+      logging:
+
+    service:
+      pipelines:
+        traces:
+          receivers: [otlp]
+          processors: []
+          exporters: [logging]
```
Comment (frzifus) [inline tests/e2e/autoscale/00-install.yaml:32]: missing CRLF

File: tests/e2e/autoscale/01-assert.yaml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+---
+apiVersion: opentelemetry.io/v1alpha1
+kind: OpenTelemetryCollector
+
+metadata:
+  name: simplest
+status:
+  scale:
+    replicas: 2
```
Comment (frzifus) [inline tests/e2e/autoscale/01-assert.yaml:18]: missing CRLF

File: tests/e2e/autoscale/01-install.yaml [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+          image: ghcr.io/open-telemetry/opentelemetry-collector-contrib/tracegen:latest
+          command:
+            - "./tracegen"
+          args:
+            - -otlp-endpoint=simplest-collector-headless:4317
+            - -otlp-insecure
+            - -duration=1m
+            - -workers=20
+      restartPolicy: Never
+  backoffLimit: 4
```
Comment (frzifus) [inline tests/e2e/autoscale/01-install.yaml:19]: missing CRLF

File: pkg/collector/upgrade/v0_56_0_test.go [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+package upgrade_test
+
+import (
+	"fmt"
+	"testing"
+)
+
+func Test0_56_0Upgrade(t *testing.T) {
+	fmt.Println("NYI")
```
Comment (frzifus) [inline pkg/collector/upgrade/v0_56_0_test.go:23]: What is `NYI`?

**Review by frzifus — COMMENTED** [review]
ive a few nit-picks :)

File: pkg/collector/upgrade/v0_56_0_test.go (hunk 11 above)
Comment (kevinearls) [reply pkg/collector/upgrade/v0_56_0_test.go]: Not yet implemented.  :-)  It is now.

File: pkg/collector/upgrade/v0_56_0.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+			updated := existing.DeepCopy()
+			updated.Spec.ScaleTargetRef = autoscalingv1.CrossVersionObjectReference{
+				Kind:       "OpenTelemetryCollector",
+				Name:       naming.OpenTelemetryCollectorName(otelcol.Name),
+				APIVersion: v1alpha1.GroupVersion.String(),
+			}
+			patch := client.MergeFrom(&existing)
+			err := u.Client.Patch(ctx, updated, patch)
+			if err != nil {
+				return nil, err
```
Comment (pavolloffay) [inline pkg/collector/upgrade/v0_56_0.go:66]: We could continue on an error and then return a multierror at the end wrapped with a message that the  upgrade of HPA was not successful. 

Either way the upgrade will have to be restarted or done manually.

File: pkg/collector/upgrade/v0_56_0.go [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+			}
+		}
+	}
+
+	u.Log.Info("in upgrade0_56_0", "Otel Instance", otelcol.Name, "Upgrade version", u.Version.String())
```
Comment (pavolloffay) [inline pkg/collector/upgrade/v0_56_0.go:71]: IIRC the other upgrade procedures use event API to record the "log" messages for the user.

File: pkg/collector/upgrade/v0_56_0.go [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+		client.MatchingLabels(map[string]string{
+			"app.kubernetes.io/instance":   fmt.Sprintf("%s.%s", otelcol.Namespace, otelcol.Name),
+			"app.kubernetes.io/managed-by": "opentelemetry-operator",
+		}),
+	}
+
+	hpaList := &autoscalingv1.HorizontalPodAutoscalerList{}
+	ctx := context.Background()
+	if err := u.Client.List(ctx, hpaList, listOptions...); err != nil {
+		return nil, err
```
Comment (pavolloffay) [inline pkg/collector/upgrade/v0_56_0.go:50]: We should wrap the error to make it clear what operation failed

File: pkg/collector/upgrade/v0_56_0_test.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+	collectorInstance.Status.Version = "0.55.0"
+	versionUpgrade := &upgrade.VersionUpgrade{
+		Log:      logger,
+		Version:  version.Get(),
+		Client:   k8sClient,
+		Recorder: record.NewFakeRecorder(upgrade.RecordBufferSize),
+	}
+	upgradedInstance, err := versionUpgrade.ManagedInstance(context.Background(), collectorInstance)
+	assert.NoError(t, err)
+	assert.Equal(t, one, *upgradedInstance.Spec.MinReplicas)
```
[the remark this answers is not in the document]
Comment (kevinearls) [reply pkg/collector/upgrade/v0_56_0_test.go:58]: I'm not sure if this is possible as upgrade get the HPA by calling the k8sclient.  It's not in the object that is passed in:

https://jenkins-cpaas-distributed-tracing.apps.cpaas-poc.r6c9.p1.openshiftapps.com/

[the remark this answers is not in the document]
Comment (pavolloffay) [reply pkg/collector/upgrade/v0_56_0_test.go:58]: It should be possible the test can access the k8sClient and create the objects before running the upgrade procedure.

**Review by pavolloffay — APPROVED** [review]
I am merging this to unblock 0.56 release. 

Consider improving the upgrade test in a follow up PR.

**Pull request merged**